turbine/src/xdp.rs

```diff
@@ -0,0 +1,183 @@
+    const DEFAULT_CHANNEL_CAP: usize = 1_000_000;
+}
+
+impl Default for XdpConfig {
+    fn default() -> Self {
+        Self {
+            interface: None,
+            cpus: vec![],
+            zero_copy: false,
+            channel_cap: Self::DEFAULT_CHANNEL_CAP,
+        }
+    }
+}
+
+impl XdpConfig {
+    pub fn new(interface: Option<impl Into<String>>, cpus: Vec<usize>, zero_copy: bool) -> Self {
+        Self {
+            interface: interface.map(|s| s.into()),
+            cpus,
+            zero_copy,
+            channel_cap: XdpConfig::DEFAULT_CHANNEL_CAP,
+        }
+    }
+}
+
+pub(crate) struct XdpSender {
+    senders: Vec<Sender<(Vec<SocketAddr>, shred::Payload)>>,
+}
+
+impl XdpSender {
+    #[inline]
+    pub(crate) fn try_send(
+        &self,
+        shred_index: u32,
+        addr: Vec<SocketAddr>,
+        payload: shred::Payload,
+    ) -> Result<(), TrySendError<(Vec<SocketAddr>, shred::Payload)>> {
+        self.senders[shred_index as usize % self.senders.len()].try_send((addr.clone(), payload))
+    }
+}
+
+pub(crate) struct XdpRetransmitter {
+    threads: Vec<thread::JoinHandle<()>>,
+}
+
+impl XdpRetransmitter {
+    #[cfg(not(target_os = "linux"))]
+    pub(crate) fn new(
+        _config: XdpConfig,
+        _src_port: u16,
+    ) -> Result<(Self, XdpSender), Box<dyn Error>> {
+        Err("XDP is only supported on Linux".into())
+    }
+
+    #[cfg(target_os = "linux")]
+    pub(crate) fn new(
+        config: XdpConfig,
+        src_port: u16,
+    ) -> Result<(Self, XdpSender), Box<dyn Error>> {
+        use caps::{
+            CapSet,
+            Capability::{CAP_BPF, CAP_NET_ADMIN, CAP_NET_RAW},
+        };
+
+        // switch to higher caps while we setup XDP. We assume that an error in
+        // this function is irrecoverable so we don't try to drop on errors.
+        for cap in [CAP_NET_ADMIN, CAP_NET_RAW, CAP_BPF] {
+            caps::raise(None, CapSet::Effective, cap)
+                .map_err(|e| format!("failed to raise {cap:?} capability: {e}"))?;
+        }
+
+        let dev = Arc::new(if let Some(interface) = config.interface {
+            NetworkDevice::new(interface).unwrap()
+        } else {
+            NetworkDevice::new_from_default_route().unwrap()
+        });
+
+        let ebpf = if config.zero_copy {
+            Some(
+                load_xdp_program(dev.if_index())
+                    .map_err(|e| format!("failed to attach xdp program: {e}"))?,
+            )
+        } else {
+            None
+        };
+
+        for cap in [CAP_NET_ADMIN, CAP_NET_RAW, CAP_BPF] {
+            caps::drop(None, CapSet::Effective, cap).unwrap();
+        }
+
+        let (senders, receivers) = (0..config.cpus.len())
+            .map(|_| crossbeam_channel::bounded(config.channel_cap))
+            .unzip::<_, _, Vec<_>, Vec<_>>();
+
+        let mut threads = vec![];
+
+        let (drop_sender, drop_receiver) = crossbeam_channel::bounded(1_000_000);
+        threads.push(
```

Comment:
1 M is too many, we would not want to have 1 GB worth of packets stuck in a buffer. Suggest cutting it down to ~1 block worth of shreds (100K). And use a const to define it of course.

Reply:
Suggesting asking more questions and making fewer statements! 😋

(We've discussed this in private for... reasons)